neps/nep-0509.md

```diff
@@ -128,6 +128,51 @@ TODO: This is essentially going to be describing the exact structure of `ChunkSt
+
+There are two types of size limits:
+* Hard limit - the size must be below this limit, anything else is considered invalid
+* Soft limit - things are added until the limit is exceeded, after that things stop being added. The last added thing is allowed to slightly exceed the limit.
+
+The limits are:
+* `max_transaction_size = 1.5 MiB`
+    * All transactions must be below 1.5 MiB, otherwise they'll be considered invalid and rejected.
+    * Previously was 4MiB, now reduced to 1.5MiB
+* `max_receipt_size - 4 MiB`:
+    * All receipts must be below 4 MiB, otherwise they'll be considered invalid and rejected.
+    * Previously there was no limit on receipt size. Set to 4MiB, might be reduced to 1.5MiB in the future to match the transaction limit.
+* `combined_transactions_size_limit - 2 MiB`
+    * Hard limit on total size of transactions from this and previous chunk. `ChunkStateWitness` contains transactions from two chunks, this limit applies to the sum of their sizes.
+* `new_transactions_validation_state_size_soft_limit - 500 KiB`
+    * Validating new transactions generates storage proof (recorded trie nodes), which has to be limited. Once transaction validation generates more storage proof than this limit, the chunk producer stops adding new transactions to the chunk.
+* `per_receipt_storage_proof_size_limit - 4 MB`
+    * Executing a receipt generates storage proof. A single receipt is allowed to generate at most 4MB of storage proof. This is a hard limit, receipts which generate more than that will fail.
+* `main_storage_proof_size_soft_limit - 3 MB`
+    * This is a limit on the total size of storage proof generated by receipts in one chunk. Once receipts generate more storage proof than this limit, the chunk producer stops processing receipts and moves the rest to the delayed queue.
+    * It's a soft limit, which means that the total size of storage proof could reach 7 MB (2.99MB + one receipt which generates 4MB of storage proof)
+    * Due to implementation details it's hard to find the exact amount of storage proof generated by a receipt, so an upper bound estimation is used instead. This upper bound assumes that every removal generates additional 2000 bytes of storage proof, so receipts which perform a lot of trie removals might be limited more than theoretically applicable.
+* `outgoing_receipts_usual_size_limit - 100 KiB`
+    * Limit on the size of outgoing receipts to another shard. Needed to keep the size of `source_receipt_proofs` small.
+    * On most block heights a shard isn't allowed to send receipts larger than 100 KiB to another shard.
+* `outgoing_receipts_big_size_limit - 4.5 MiB`
+    * On every block height there's one special "allowed shard" which is allowed to send larger receipts, up to 4.5 MiB in total.
+    * A receiving shard will receive receipts from `num_shards - 1` shards using the usual limit and one shard using the big limit.
```

Comment:
How is this shard determined? Is it going to be randomly picked every block? Is it going to be a gateway shard that allows for big receipts?

Reply:
It's the `allowed_shard` from congestion control NEP. I'll add a paragraph about it.